sw/device/tests/usbdev_vbus_test.c

```diff
@@ -0,0 +1,116 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// USB VBUS test
+//
+// Checks for the presence of VBUS from the host/DPI model, optionally checking
+// that VBUS is low initially when the test starts.
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/status.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_usbdev.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#define USBDEV_BASE_ADDR TOP_EARLGREY_USBDEV_BASE_ADDR
+
+/**
+ * Are we expecting VBUS to be low at the start of the test, ie. no connection
+ * to the physical host or the DPI model is inactive?
+ * Typically useful only for an FPGA target that is being manually supervised.
+ */
+static const bool kCheckLowFirst = false;
+
+/**
+ * USB device handle
+ */
+static dif_usbdev_t usbdev;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+// Wait with timeout until the VBUS/SENSE signal is in the expected state.
+static status_t vbus_wait(bool expected, uint32_t timeout_micros) {
+  ibex_timeout_t timeout = ibex_timeout_init(timeout_micros);
+  do {
+    // Read the current state of VBUS.
+    bool vbus;
+    TRY(dif_usbdev_status_get_sense(&usbdev, &vbus));
+    if (vbus == expected) {
+      return OK_STATUS();
+    }
+  } while (!ibex_timeout_check(&timeout));
+
+  return DEADLINE_EXCEEDED();
+}
+
+OTTF_DEFINE_TEST_CONFIG();
+
+bool test_main(void) {
+  CHECK(kDeviceType == kDeviceSimDV || kDeviceType == kDeviceSimVerilator ||
+            kDeviceType == kDeviceFpgaCw310,
+        "This test is not expected to run on platforms other than the "
+        "Verilator/DV simulation or CW310 FPGA. It needs either the DPI model "
+        "or a VBUS/SENSE connection to a physical host");
```

Comment:
I feel a little iffy about having this in the C code. This blurb may add maintenance burdens that the build system should already handle, and it makes us have to modify yet another place to add support for new platforms.

Relatedly, I wonder what `kDeviceType` would be for the SiVal platform.

Reply:
Fair point. I've just been mimicking what was there in usbdev_test and granted we should have CW340 support soon, at which point I'll just do a wholesale change to all usbdev-related tests at the same time, once they're working. I guess the same happens for SiVal.

I understood the purpose of the CHECK here was to stop dead when someone attempts the first run on a new platform, as a warning that it hasn't been written/tested with that in mind.

Reply:
I've removed this check now, having recently discovered in the CW340 bring up that the usbdev_test code no longer contains a check. When that became usable in ASIC simulation with EDA tools too, all targets were covered. That's more true of this simple VBUS test than any other.
Thanks.